lib/play_game.py

```diff
@@ -3,27 +3,38 @@
 
 
 class PlayGame:
+
+    current_player = "X"
+    next_player = "O"
+
     def __init__(self, board, ui):
         self.board = board
         self.ui = ui
 
     def start(self):
         self.ui.display("Welcome to TIC TAC TOE!")
         self.board.display_board()
+        self.ui.display(self.current_player + " goes first")
         while not self.board.board_full(self.board.board):
-            self.turn()
+            self.turn(self.current_player)
+            self.switch_players()
 
-    def turn(self):
+    def turn(self, mark):
```

Comment:
I also am not sure if anyone mentioned naming conventions for functions and classes yet. 

Typically your function names should be expressive of what the function is doing. In this case, `turn` would be more clear as `take_turn`. Usually the function names are going to be a verb, since they are performing an action.

The class name, on the other hand, is the representation for the object, so a noun makes more sense. So instead of `PlayGame`, which sounds more like a function name, I would just call this class `Game`.

Reply:
So for this function, you have a couple of different things you are checking

1. Is the input valid?
2. Is the position taken?

The response to either of those things could be true or false, so your tests should reflect the different combinations of those results. So you would have four test cases:

1. the input is valid and the position is taken
2. the input is valid and the position is not taken
3. the input is not valid and the position is taken
4. the input is not valid and the position is not taken

You would also be mocking your ui methods here, validating that they were called. Your `board.display_board` method should also be calling the ui method. The only place you should be calling `print` or `input` should be in that `UI` class.

Reply:
The thing I cannot wrap
 my head around it is how to get around the recursiveness of the turn method?....as long as I call the method I've got this error in the test "Recursion detected "

Reply:
Thank you for the hint!